File: cpp/misc/move_zero.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

/*
    Move the zeros to the end,
    If zeros to move to the start,
    we should start from the end of the
    array for efficiency
*/


void move_zero(vector<int>& nums)
{
    size_t n = nums.size();
    if (n < 2) {
        return;
    }
    
    int pos = 0, start_zero = -1;
    while (pos < n) {
        if (nums[pos] > 0 && start_zero != -1) {
            nums[start_zero] = nums[pos]; 
            nums[pos] = 0;
            start_zero++;
        } else {
            if (start_zero == -1) {
                start_zero = pos;
            }
        }
        pos++;
    }
}

int main() {
    vector<int> x = {0, 1, 2,0, 0, 3,0,4};
    move_zero(x);
    for_each(x.begin(), x.end(), [](int a) { cout << a << " "; });
    cout << endl;
}

File: python/misc/move_zero.py
'''
Move the zeros to the end,
If zeros to move to the start,
we should start from the end of the
array for efficiency

'''

def move_zero(arr):
    if len(arr) < 2:
        return

    i = len(arr)
    j = 0
    start_zero = -1
    while j < i:
        if arr[j] != 0:
            if start_zero != -1:
                arr[start_zero] = arr[j]
                arr[j] = 0
                start_zero += 1
        else:
            if start_zero == -1:
                start_zero = j
        j += 1

def move_zero_first(arr):
    if len(arr) < 2:
        return

    i = len(arr)
    j = i - 1
    start_zero = -1
    while j > -1:
        if arr[j] != 0:
            if start_zero != -1:
                arr[start_zero] = arr[j]
                arr[j] = 0
                start_zero -= 1
        else:
            if start_zero == -1:
                start_zero = j
        j -= 1

x = [0,0,1,2, 0, 4, 5, 0, 6, 7, 0]
y = [0,0,1,2, 0, 4, 5, 0, 6, 7, 0]
move_zero(x)
print(x)
move_zero_first(y)
print(y)


